src/hooks/use-web3-call-state.ts

```diff
@@ -79,44 +79,69 @@ export function useWeb3CallState(
     });
   };
 
-  if (web3Account === "ETH") {
-    window.ethereum?.on("accountsChanged", (accounts: string[]) => {
-      console.log("!!! ETH accountsChanged", accounts);
-      setWeb3Account("ETH");
-      setWeb3Address(accounts[0], "accountsChanged");
-    });
-  }
+  useEffect(() => {
+    if (window.ethereum && web3Account === "ETH") {
+      const handleAccountsChanged = (accounts: string[]) => {
+        console.log("!!! ETH accountsChanged", accounts);
+        if (!(web3Account === "ETH" && web3Address === accounts[0])) {
+          setWeb3Account("ETH");
+          setWeb3Address(accounts[0], "accountsChanged");
+        }
+      };
+      window.ethereum.on("accountsChanged", handleAccountsChanged);
+      return () => {
+        window.ethereum.removeListener(
+          "accountsChanged",
+          handleAccountsChanged,
+        );
+      };
+    }
+  });
```

Comment:
A `useEffect` without a dependency array fires on every re-render of the component. See https://react.dev/reference/react/useEffect:

> If you omit this argument, your Effect will re-run after every re-render of the component.

Is that really what you intended?  It's normally a mistake and gets flagged as such by eslint-plugin-react-hooks.  I wonder in this case whether actually your dependency array should be `[window.ethereum]` because only if that changes do you need this code to re-execute. But that might not work if `window.ethereum` is not guaranteed to be a stable reference.

I was about to suggest it's therefore clearer to remove the `useEffect` entirely, but in fact you're making (good and appropriate) use of the cleanup function. 

I don't think this is a blocker as in the worse case we attach and re-attach the listener on every render, which is unlikely to be a concern.

Similar comment applies to the other useEffects.

Reply:
OK, so for this one I want `[web3Account]` and for all the others I want the empty array `[]` as they only need to run once. Is that correct @tackley?

I `useEffect` because before I noticed multiple event handlers being added and never removed, and this allows me to set a cleanup function. I don't know about `window.ethereum` being used as a React dependency but it shouldn't need to be there... I can't see why it would suddenly change from falsey to truthy.

Reply:
Yes, that looks great. And agree with your conclusions there, I think what you've updated the dependency arrays to looks good.